chainercv/links/model/pspnet/pspnet.py

```diff
@@ -0,0 +1,476 @@
+                self.out_aux = L.Convolution2D(
+                    512, n_class, 3, 1, 1, False, initialW)
+
+            # Main branch
+            feat_size = (input_size[0] // 8, input_size[1] // 8)
+            self.ppm = PyramidPoolingModule(2048, feat_size, pyramids)
+            self.cbr_main = ConvBNReLU(4096, 512, 3, 1, 1)
+            self.out_main = L.Convolution2D(
+                512, n_class, 1, 1, 0, False, initialW)
+
+        self.mean = mean
+
+        if pretrained_model in self._models:
+            path = download_model(self._models[pretrained_model]['url'])
+            chainer.serializers.load_npz(path, self)
+            self._use_pretrained_model = True
+            print('Pre-trained model has been loaded:', pretrained_model)
+        elif pretrained_model:
+            self._use_pretrained_model = False
+            chainer.serializers.load_npz(pretrained_model, self)
+            print('Pre-trained model has been loaded:', pretrained_model)
+        else:
+            self._use_pretrained_model = False
+
+    @property
+    def n_class(self):
+        return self.out_main.out_channels
+
+    def __call__(self, x):
+        """Forward computation of PSPNet
+
+        Args:
+            x: Input array or Variable.
+
+        Returns:
+            Training time: it returns the outputs from auxiliary branch and the
+                main branch. So the returned value is a tuple of two Variables.
+            Inference time: it returns the output of the main branch. So the
+                returned value is a sinle Variable which forms
+                ``(N, n_class, H, W)`` where ``N`` is the batchsize and
+                ``n_class`` is the number of classes specified in the
+                constructor. ``H, W`` is the input image size.
+
+        """
+        if chainer.config.train:
+            aux, h = self.trunk(x)
+            aux = F.dropout(self.cbr_aux(aux), ratio=0.1)
+            aux = self.out_aux(aux)
+            aux = F.resize_images(aux, x.shape[2:])
+        else:
+            h = self.trunk(x)
+
+        h = self.ppm(h)
+        h = F.dropout(self.cbr_main(h), ratio=0.1)
+        h = self.out_main(h)
+        h = F.resize_images(h, x.shape[2:])
+
+        if chainer.config.train:
+            return aux, h
+        else:
+            return h
+
+    def prepare(self, img):
+        """Preprocess an image for feature extraction.
+
+        The image is subtracted by a mean image value :obj:`self.mean`.
+
+        Args:
+            img (~numpy.ndarray): An image. This is in CHW and RGB format.
+                The range of its value is :math:`[0, 255]`.
+
+        Returns:
+            ~numpy.ndarray:
+            A preprocessed image.
+
+        """
+        if self.mean is not None:
+            img -= self.mean[:, None, None]
+            img = img.astype(np.float32, copy=False)
+            if self._use_pretrained_model:
+                # Pre-trained model is trained for BGR images
```

Comment:
Can you flip the order of in_channels of the weights in conversion code?
By doing so, this if statement can be deleted.
I want to make `pspnet.py` to be as simple as possible at the expense of more complex conversion code because people would rarely see the conversion code.

Reply:
OK, I'll try to do it.